crates/text/src/lib.rs

```diff
@@ -0,0 +1,365 @@
+}
+
+impl<'a> From<Vec<Text<'a>>> for Text<'a> {
+    fn from(extra: Vec<Text<'a>>) -> Self {
+        let mut text = Text::of("");
+        text.extra = extra;
+        text
+    }
+}
+
+// More and better tests
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+
+    use super::*;
+
+    #[test]
+    fn simple() {
+        assert_eq!(
+            json!({
+                "text": "Some basic text, with no formatting at all!",
+            }),
+            serde_json::to_value(Text::of("Some basic text, with no formatting at all!")).unwrap()
+        );
+    }
+
+    #[test]
+    fn with_color() {
+        Text::of("Red colored text!").red();
+        Text::of("Red colored text!").color(Color::Red);
+    }
+
+    #[test]
+    fn with_single_style() {}
+
+    #[test]
+    fn multiple_styles() {}
+
+    #[test]
+    fn complex() {
+        let complex_json = json!({
+            "text": "Complex text",
+            "extra": [{
+                "text": "Un formatted extra",
+            }, {
+                "translate": "Formatted translation",
+                "with": [],
+                "color": "gold",
+                "bold": true,
+                "italic": false,
+            }],
+            "italic": true,
+            "clickEvent": {
+                "action": "open_url",
+                "value": "https://feather.rs",
+            }
+        });
+
+        let complex_text = Text::of("Complex text")
+            .extra(Text::of("Un formatted extra"))
+            .extra(
+                Text::translation("Formatted translation", vec![])
+                    .gold()
+                    .bold()
+                    .not_italic(),
+            )
+            .italic()
+            .on_click_open_url("https://feather.rs");
+
+        assert_eq!(complex_json, serde_json::to_value(complex_text).unwrap())
+    }
+
+    #[test]
+    fn custom_deserializing() {
+        let json_value = json!({
+            "text": "Hello",
+            "extra": [" ", "world"],
+        });
+
+        let json_text: Text = serde_json::from_value(json_value.clone()).unwrap();
+
+        let text = Text::of("Hello")
+            .extra(Text::of(" "))
+            .extra(Text::of("world"));
+
+        assert_eq!(json_text, text);
+
+        let text_value = serde_json::to_value(text).unwrap();
+        assert_eq!(json_value, text_value);
+    }
+
+    #[test]
+    fn custom_serializing() {
+        assert_eq!(
+            json!({
+                "text": "",
+                "extra": vec!["Hello", " ", "world"],
+            }),
+            serde_json::to_value(Text::from(vec![
+                Text::of("Hello").red(),
+                Text::of(" "),
+                Text::of("world").blue(),
+            ]))
+            .unwrap(),
+        );
+    }
+
+    #[test]
+    fn json() {
+        let json = serde_json::to_string_pretty(&Text::of("hello world").red().bold()).unwrap();
+        println!("{}", json);
+    }
+}
```

Comment:
Some test cases to consider:
* Just a JSON array - valid according to the wiki (empty root node with children from the array)
* Raw JSON primitives that aren't strings, like an integer